packages/vm/src/context.rs

```diff
@@ -62,229 +58,243 @@ impl GasState {
-    let data = ContextData::<S, Q> {
-        gas_state: GasState::with_limit(gas_limit),
-        storage: None,
-        storage_readonly: true,
-        querier: None,
-        wasmer_instance: None,
-    };
-    let heap_data = Box::new(data); // move from stack to heap
-    Box::into_raw(heap_data) as *mut c_void // give up ownership
+impl<S: Storage, Q: Querier> WasmerEnv for Env<S, Q> {
+    fn init_with_instance(&mut self, _instance: &WasmerInstance) -> Result<(), HostEnvInitError> {
+        Ok(())
+    }
 }
 
-fn destroy_unmanaged_context_data<S: Storage, Q: Querier>(ptr: *mut c_void) {
-    if !ptr.is_null() {
-        // obtain ownership and drop instance of ContextData when box gets out of scope
-        let _dying = unsafe { Box::from_raw(ptr as *mut ContextData<S, Q>) };
+impl<S: Storage, Q: Querier> Env<S, Q> {
+    pub fn new(gas_limit: u64) -> Self {
+        Env {
+            data: Arc::new(RwLock::new(ContextData::new(gas_limit))),
+        }
     }
-}
 
-/// Get a mutable reference to the context's data. Ownership remains in the Context.
-fn get_context_data_mut<'a, 'b, S: Storage, Q: Querier>(
-    ctx: &'a mut Ctx,
-) -> &'b mut ContextData<S, Q> {
-    unsafe {
-        let ptr = ctx.data as *mut ContextData<S, Q>;
-        ptr.as_mut()
-            .expect("The pointer ctx.data was null in get_context_data_mut; this is a bug.")
+    pub fn with_context_data_mut<Callback, CallbackReturn>(
+        &self,
+        callback: Callback,
+    ) -> CallbackReturn
+    where
+        Callback: FnOnce(&mut ContextData<S, Q>) -> CallbackReturn,
+    {
+        let mut guard = self.data.as_ref().write().unwrap();
+        let context_data = guard.borrow_mut();
+        callback(context_data)
+    }
+
+    pub fn with_context_data<Callback, CallbackReturn>(&self, callback: Callback) -> CallbackReturn
+    where
+        Callback: FnOnce(&ContextData<S, Q>) -> CallbackReturn,
+    {
+        let guard = self.data.as_ref().read().unwrap();
+        let context_data = guard.borrow();
+        callback(context_data)
+    }
+
+    pub fn with_gas_state_mut<Callback, CallbackReturn>(&self, callback: Callback) -> CallbackReturn
+    where
+        Callback: FnOnce(&mut GasState) -> CallbackReturn,
+    {
+        self.with_context_data_mut(|context_data| callback(&mut context_data.gas_state))
+    }
+
+    pub fn with_gas_state<Callback, CallbackReturn>(&self, callback: Callback) -> CallbackReturn
+    where
+        Callback: FnOnce(&GasState) -> CallbackReturn,
+    {
+        self.with_context_data(|context_data| callback(&context_data.gas_state))
     }
-}
 
-fn get_context_data<'a, 'b, S: Storage, Q: Querier>(ctx: &'a Ctx) -> &'b ContextData<S, Q> {
-    unsafe {
-        let ptr = ctx.data as *mut ContextData<S, Q>;
-        ptr.as_ref()
-            .expect("The pointer ctx.data was null in get_context_data; this is a bug.")
+    pub fn with_func_from_context<Callback, CallbackData>(
```

Comment:
Crazy, so this is how you can call into the wasmer instance from within a host function?

It took me a bit to decipher what it did, but it makes good sense